### ISSUE TITLE
Questions about queue ordering in the SYCL backend

### ISSUE THREAD

**Issue by mrnorman** [mrnorman/YAKL#74]
Why do we have to explicitly tell parallel_for to synchronize on the `copyEvent` here? Since the `sycl::queue` is declared to be in order, shouldn't this dependency already be respected?

https://github.com/mrnorman/YAKL/blob/d19dd2993c6e96bcbb259d75392fe91325daabb2/src/YAKL_parallel_for_common.h#L237

I have a similar question for why the host has to synchronize on the kernel event here:
https://github.com/mrnorman/YAKL/blob/d19dd2993c6e96bcbb259d75392fe91325daabb2/src/YAKL_parallel_for_common.h#L240

Without asking the host to wait on the kernel, and without explicitly asking the kernel to wait on the copy event, we get wrong answers or hangs. CUDA and HIP seem to handle this situation fine, so is there something about `sycl::queue` ordering semantics I'm not understanding? Thanks for the help.

**Comment by abagusetty** [comment]
@mrnorman There is absolutely no need for the `copyEvent` to be passed to the next kernel launch in case of the sycl::queue of in_order. I was just verifying this to be removed on the new SDK, hardware as I type this. It was added with an intention to make it bullet proof when the runtime was not so mature. 

Regarding the host-wait, I will have to verify what is the behavior of `free_device()` when the kernel is still in works. 
```
https://github.com/mrnorman/YAKL/blob/d19dd2993c6e96bcbb259d75392fe91325daabb2/src/YAKL_parallel_for_common.h#L266
```

In theory, for an in_order queue which is similar to CUDA/HIP stream FIFO, all the synchronization and dependencies needs to be the same. I will update you here today on the latest SDK and if there is a need to update the logic

**Comment by mrnorman** [comment]
OK, sweet, I confirmed it's fine to remove the copyEvent from the kernel launch. So now it's just down to the host wait. 

If the memory pool is enabled (which it is by default), free_device tell the pool to release that entry and make it available for future use. So it will be aliased by other things on the device while the kernel is still running, but nothing will be *done* to it until the kernel finishes because all future work is in the same in-order queue.

There *is* an issue with streams currently that needs to be handled if the pool is enabled, but if the user doesn't use streams, I believe the kernel.wait() shouldn't be necessary right now.

Currently, if I remove the `kernelEvent.wait()` line, I get a hang in GPU tests.

**Comment by mrnorman** [comment]
I just pushed the removal of copyEvent to main branch

**Comment by abagusetty** [comment]
Update: I was able to reproduce the hang with the tests or incorrect results (in some rare cases) when removing the two concerned instance of `kernelEvent.wait()`. 

The above can be resolved with `ZE_SERIALIZE=2` environment variable. This is just a work-around and I am looking into for a more permanent solution. Solution: It should be safe to remove the two instances of `.wait()` by design to align similarly with CUDA and HIP.

**Comment by mrnorman** [comment]
Clarifying two things Abhishek said on a call recently:
1. Our current hypothesis is that the problem lies in the SYCL runtime. This is a hypothesis and not a guarantee that YAKL is free of user programming bugs in the SYCL backend. The reasoning for the hypothesis is the same behavior works for CUDA and HIP; and that `ZE_SERIALIZE=2` fixes the problem.
2. This only happens when the pool allocator is turned **on**. With `export GATOR_DISABLE=1`, the problem goes away. YAKL knowingly aliases pointers in the pool at the same point in time from the host's perspective. However, the device has everything running in the same stream, and the aliased pointers are therefore guaranteed to never be used at the same time on the device.

**Comment by mrnorman** [comment]
@abagusetty , it occurs to me there is likely a workaround in the meantime for SYCL. We may be able to use YAKL's "free_with_event_dependencies" routines to get rid of the `.wait()`, but let's see what shakes out of investigations into SYCL's runtime behavior first.

**Comment by omarkahmed** [comment]
I confirm that I am getting the error you are seeing @abagusetty & @mrnorman with the 9/30 nightly compiler and 10/06 MKL:

**Comment by omarkahmed** [comment]
Looks like I'm getting the same issue with the 10/7 compiler nightly (same version of MKL, which was built with the 9/30 nightly).  I can go ahead and file a JIRA for this issue.  Please let me know if the FFT unit test is the simplest representation of this issue (this is the one that was most reliable for me).

**Comment by omarkahmed** [comment]
The issue also reproducing with 10/14 nightly, and I'm seeing it on the CArray unit test.  I'm attaching a stripped down version of the CArray test which also is demonstrating the issue.

[CArray.cpp.txt](https://github.com/mrnorman/YAKL/files/9804317/CArray.cpp.txt)

Will file a JIRA based on this reproducer.

**Comment by omarkahmed** [comment]
Looks like the hangs are related to exit(-1) being called prior to yakl::finalize() in the die() functions.

**Comment by mrnorman** [comment]
@omarkahmed , I've pushed a commit to `main` that uses `yakl_throw` instead of `exit(-1)`. So on current main, this issue presents as wrong answer bugs rather than hangs.

**Comment by mrnorman** [comment]
As another data point, I implemented the is_device_coypable type traits for functors small enough to launch directly without manually copying. For these situations, I am able to remove `wait()` after kernel launches. This is currently on main branch after commit 041115f85a9df912830134cf996aa7162fee0b39 . So this issue seems to have to do specifically with the copying of functors and maintaining expected queue ordering.

**Comment by mrnorman** [comment]
I've improved things a bit in commit a89671954b03c779e0c0ce6fee47cac68ef64db0 to only wait on the copy event rather than the kernel event for functors too large to use is_device_copyable.

We pass a reference of the functor to `parallel_for` and then copy the memory from that address to the device. I wonder if the host is overwriting the memory at that location before or during the host-to-device copy? I'll need to think some more on what the semantics should be expected to be for that. For CUDA, this pattern works fine, at least in unit tests. I've tried switching to always using the explicit functor copying in CUDA, and things still work. So it may be that C++ semantics don't specify what the permanence of lambdas should be in host memory; and Nvidia behaves one way while Intel behaves another.

I'd like to know your thoughts on this.

Regardless, improvement **should** be greatly improved for most kernels with the behavior on current main.

**Comment by omarkahmed** [comment]
@mrnorman , I am seeing much better performance on SAMXX 1 moment.  For time steps > 1, looking at NCRMS=224, 448, 896, the wall time goes down by about a factor of 2-3 (with the bigger differences at smaller NCRMS).  Looks like the first time step is still very expensive on the PVC, however.   I went ahead and filed a JIRA to try to understand the reason for the additional waits.

**Comment by mrnorman** [comment]
I'll reiterate the data point that we only get wrong answers with the pool turned on, which might invalidate the previous hypothesis. I need to think on this.
